### ISSUE TITLE
Report inconsistency in the command `JSON.SET`

### ISSUE THREAD

**Issue by arpitbbhayani** [DiceDB/dice#381]
This issue is all about ensuring we are as close to Redis as possible. The command in focus for this issue is `JSON.SET`.

Go through the [official documentation](https://redis.io/docs/latest/commands) of the command `JSON.SET` on Redis and identify the inconsistencies. The inconsistencies could be in

1. unhandled edge case
1. unexpected behavior
1. unsupported option

Because we are trying to be compatible with Redis v7.2.5, I would recommend you try out different variants of the command with different inputs on that specific version. The instructions on running Redis v7.2.5 locally

- [from source code](https://gist.github.com/arpitbbhayani/94aedf279349303ed7394197976b6843)
- [use Docker](https://hub.docker.com/_/redis)

Once you find the discrepancy, you can either

1. raise an issue on [Dice repository](https://github.com/dicedb/dice) with details, or
1. try to fix it yourself and raise a PR

If you are [raising the issue](https://github.com/DiceDB/dice/issues/new?assignees=&labels=&projects=&template=inconsistent_dicedb_vs_redis.md&title=Inconsistent+%60%7BCMD%7D%60%3A+%3CDescribe+the+error+in+one+concise+line%3E), make sure you provide the details such as

0. [use the template](https://github.com/DiceDB/dice/issues/new?assignees=&labels=&projects=&template=inconsistent_dicedb_vs_redis.md&title=Inconsistent+%60%7BCMD%7D%60%3A+%3CDescribe+the+error+in+one+concise+line%3E) and provide the following details
1. steps to reproduce (series of commands)
1. observed output on DiceDB
1. observed output on Redis v7.2.5

Also, feel free to update the documentation and raise the PR in the [docs repository](https://github.com/dicedb/docs).

> You will need to go deeper into the command make sure you are covering all cases and reporting the inconsistencies or fixing them. The deeper the work, the better our stability will be. Also, it is possible that we do not find any discrepancies, so please mention the same in the comment on this issue. Mention the PR or issue links that you create under this issue.

**Comment by kushal0511-not** [comment]
@arpitbbhayani can i take this up?

**Comment by arpitbbhayani** [comment]
@kushal0511-not go for it :)

**Comment by kushal0511-not** [comment]
We have some inconsistency in command due to  how we treat **JSONPath** in JSON.SET command.
here are few examples.
1 )
dice : 
![Screenshot from 2024-08-28 17-51-32](https://github.com/user-attachments/assets/51337cf0-54a3-4ad9-9c0a-90b657f1fd88)
Redis 7.2.5 :
![Screenshot from 2024-08-28 17-51-58](https://github.com/user-attachments/assets/2c0bc604-7df9-4e77-82d6-4ee3b34766f7)



2 )
dice :
![Screenshot from 2024-08-28 17-54-26](https://github.com/user-attachments/assets/e81552bf-6ac0-4d53-9ad6-205b1f8107f5)
Redis 7.2.5 :
![Screenshot from 2024-08-28 17-54-10](https://github.com/user-attachments/assets/21a6ee09-c0c6-405d-bf1a-1675469b9b0f)

cc @JyotinderSingh

**Comment by kushal0511-not** [comment]
I have tried other libraries for jsonpath, but the current library which we are using is the  closest to redis(7.2.5) jsonpath.
As we have inconsistency with the redis (like this library parse some json paths differently than redis 7.2.5) , what should be the next steps for this issue.
Should we fork this library and make changes to it 
OR
Should we create our own from scratch.


@JyotinderSingh @soumya-codes

**Comment by JyotinderSingh** [comment]
> I have tried other libraries for jsonpath, but the current library which we are using is the  closest to redis(7.2.5) jsonpath.
> As we have inconsistency with the redis (like this library parse some json paths differently than redis 7.2.5) , what should be the next steps for this issue.
> This will affect  almost all JSON related commands.
> Should we fork this library and make changes to it 
> OR
> Should we create our own from scratch.
> 
> 
> @JyotinderSingh @soumya-codes 

Thanks for looking into this @kushal0511-not, let's keep this issue open for now, we can look into fixing it at a later stage.
The partial support for the syntax is enough for the existing use cases.

**Comment by apoorvyadav1111** [comment]
Hi, I have found another inconsistency for `JSON.SET` when we try to create a string at the root path. 
dice:
![image](https://github.com/user-attachments/assets/c003459d-2b2e-4ac2-a2af-351a2bdfa2d7)
redis:
![image](https://github.com/user-attachments/assets/ee1a840a-c093-47e7-a9f8-b70ce7df5308)

I am working on `JSON.STRLEN` and found this deviation. Since this is an open issue, I feel reporting it here is better than raising another issue.
Regards,
Apoorv

**Comment by arpitbbhayani** [comment]
Hello @kushal0511-not,

There has been no activity on this issue for the past 5 days.
It would be awesome if you keep posting updates to this issue so that we know you are actively working on it.

We are really eager to close this issue at the earliest, hence if we continue to see the inactivity, we will have to reassign the issue to someone else. We are doing this to ensure that the project maintains its momentum and others are not blocked on this work.

Just drop a comment with the current status of the work or share any issues you are facing. We can always chip in to help you out.

Thanks again.

**Comment by kushal0511-not** [comment]
Hi @arpitbbhayani , i have described root cause and discussed with @JyotinderSingh also in previous comments.

**Comment by arpitbbhayani** [comment]
Hello @kushal0511-not,

There has been no activity on this issue for the past 5 days.
It would be awesome if you keep posting updates to this issue so that we know you are actively working on it.

We are really eager to close this issue at the earliest, hence if we continue to see the inactivity, we will have to reassign the issue to someone else. We are doing this to ensure that the project maintains its momentum and others are not blocked on this work.

Just drop a comment with the current status of the work or share any issues you are facing. We can always chip in to help you out.

Thanks again.

**Comment by arpitbbhayani** [comment]
Hello @kushal0511-not,

There has been no activity on this issue for the past 5 days.
It would be awesome if you keep posting updates to this issue so that we know you are actively working on it.

We are really eager to close this issue at the earliest, hence if we continue to see the inactivity, we will have to reassign the issue to someone else. We are doing this to ensure that the project maintains its momentum and others are not blocked on this work.

Just drop a comment with the current status of the work or share any issues you are facing. We can always chip in to help you out.

Thanks again.

**Comment by apoorvyadav1111** [comment]
Hi, I have been working on documentation for JSON.SET commands and have found many inconsistencies with it. This issue seems like a parent issue for the documentation. I will not be raising another issue. 

Dependent issue: #795

**Comment by arpitbbhayani** [comment]
Hello @kushal0511-not,

There has been no activity on this issue for the past 5 days.
It would be awesome if you keep posting updates to this issue so that we know you are actively working on it.

We are really eager to close this issue at the earliest, hence if we continue to see the inactivity, we will have to reassign the issue to someone else. We are doing this to ensure that the project maintains its momentum and others are not blocked on this work.

Just drop a comment with the current status of the work or share any issues you are facing. We can always chip in to help you out.

Thanks again.
